hedera-node/src/main/java/com/hedera/services/txns/token/TokenUpdateTransitionLogic.java

```diff
@@ -65,125 +67,195 @@
+				: null;
+	}
 
-		outcome = autoRenewAttachmentCheck(op, token);
-		if (outcome != OK) {
-			txnCtx.setStatus(outcome);
-			return;
+	private void throwIfAssociationIsMissingBetween(final Account newTreasury, final Token token) {
+		final var associationDoesNotExist = !newTreasury.getAssociatedTokens().contains(token.getId());
+		validateFalse(associationDoesNotExist, INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
+	}
+
+	private void throwIfCurrentTreasuryOwnsNfts(final Token token) {
+		final var treasuryIsNotEligibleToHoldNfts =
+				!allowChangedTreasuryToOwnNfts && token.getType() == NON_FUNGIBLE_UNIQUE;
+
+		if (treasuryIsNotEligibleToHoldNfts) {
+			final var tokenRelationship = tokenStore.loadTokenRelationship(token, token.getTreasury());
+			final var currentTreasuryOwnsNfts = tokenRelationship.getBalance() > 0L;
+			validateFalse(currentTreasuryOwnsNfts, CURRENT_TREASURY_STILL_OWNS_NFTS);
 		}
+	}
 
-		Optional<AccountID> replacedTreasury = Optional.empty();
-		if (op.hasTreasury()) {
-			var newTreasury = op.getTreasury();
-			if (ledger.isDetached(newTreasury)) {
-				txnCtx.setStatus(ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
-				return;
-			}
-			if (!store.associationExists(newTreasury, id)) {
-				txnCtx.setStatus(INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
-				return;
-			}
-			var existingTreasury = token.treasury().toGrpcAccountId();
-			if (!allowChangedTreasuryToOwnNfts && token.tokenType() == NON_FUNGIBLE_UNIQUE) {
-				var existingTreasuryBalance = ledger.getTokenBalance(existingTreasury, id);
-				if (existingTreasuryBalance > 0L) {
-					abortWith(CURRENT_TREASURY_STILL_OWNS_NFTS);
-					return;
-				}
-			}
-			if (!newTreasury.equals(existingTreasury)) {
-				if (ledger.isDetached(existingTreasury)) {
-					txnCtx.setStatus(ACCOUNT_EXPIRED_AND_PENDING_REMOVAL);
-					return;
-				}
-				outcome = prepNewTreasury(id, token, newTreasury);
-				if (outcome != OK) {
-					abortWith(outcome);
-					return;
-				}
-				replacedTreasury = Optional.of(token.treasury().toGrpcAccountId());
-			}
+	private void prepareNewTreasury(final TokenRelationship newTreasuryRel, final Token token) {
+		if (token.hasFreezeKey()) {
+			newTreasuryRel.changeFrozenState(false);
+		}
+		if (token.hasKycKey()) {
+			newTreasuryRel.changeKycState(true);
 		}
 
-		outcome = store.update(op, txnCtx.consensusTime().getEpochSecond());
-		if (outcome == OK && replacedTreasury.isPresent()) {
-			final var oldTreasury = replacedTreasury.get();
-			long replacedTreasuryBalance = ledger.getTokenBalance(oldTreasury, id);
-			if (replacedTreasuryBalance > 0) {
-				if (token.tokenType().equals(TokenType.FUNGIBLE_COMMON)) {
-					outcome = ledger.doTokenTransfer(
-							id,
-							oldTreasury,
-							op.getTreasury(),
-							replacedTreasuryBalance);
+		tokenStore.persistTokenRelationships(List.of(newTreasuryRel));
+	}
+
+	private void persistTransferOfTreasuryBalanceIfAppropriate(
```

Comment:
I think that it is better to use the refactored way of transferring tokens which is the one used in the `CryptoTransferTransitionLogic` -> `List<BalanceChanges>` and do `ledger.doZeroSum`